# 1. Instale la librería scapy: https://scapy.net/
2. Tutorial de uso:
https://guedou.github.io/talks/2022_GreHack/Scapy%20in%200x30%20minutes.slides.html

# 3. Verifique que la herramienta fue correctamente instalada:

a. Capture 25 paquetes de su red doméstica (no de la red de la UVG) y asígnelos a una
variable.

b. Imprima el tipo de variable, la longitud y el contenido de la variable.

c. Imprima el tipo de dato del primer paquete capturado.

d. Imprima el contenido de 5 paquetes

a. Capture 25 paquetes de su red doméstica (no de la red de la UVG) y asígnelos a una
variable

In [ ]:
from scapy.all import sniff

# Captura 25 paquetes desde una red doméstica
paquetes = sniff(count=25)




b. Imprima el tipo de variable, la longitud y el contenido de la variable

In [4]:
print("Tipo de variable:", type(paquetes))
print("Cantidad de paquetes capturados:", len(paquetes))
print("Contenido de la variable:")
print(paquetes)


Tipo de variable: <class 'scapy.plist.PacketList'>
Cantidad de paquetes capturados: 24
Contenido de la variable:
<Sniffed: TCP:24 UDP:0 ICMP:0 Other:0>


c. Imprima el tipo de dato del primer paquete capturado.

In [5]:
print("Tipo de dato del primer paquete:", type(paquetes[0]))


Tipo de dato del primer paquete: <class 'scapy.layers.l2.Ether'>


d. Imprima el contenido de 5 paquetes

In [6]:
print("Contenido de los primeros 5 paquetes:")
for i in range(5):
    print(f"Paquete {i+1}:")
    print(paquetes[i])
    print("-" * 40)


Contenido de los primeros 5 paquetes:
Paquete 1:
Ether / IP / TCP 172.64.150.27:https > 192.168.68.62:51014 PA / Raw
----------------------------------------
Paquete 2:
Ether / IP / TCP 172.64.150.27:https > 192.168.68.62:51014 PA / Raw
----------------------------------------
Paquete 3:
Ether / IP / TCP 192.168.68.62:51014 > 172.64.150.27:https A
----------------------------------------
Paquete 4:
Ether / IP / TCP 192.168.68.62:51014 > 172.64.150.27:https PA / Raw
----------------------------------------
Paquete 5:
Ether / IP / TCP 172.64.150.27:https > 192.168.68.62:51014 A
----------------------------------------


Análisis estadístico
1. Descargue e archivo analisis_paquetes.pcap y asignelo a una variable.
2. Convierta la variable a un DataFrame.
3. Muestre el contenido de las primeras 5 filas del dataset.
4. Muestre los valores de las columnas: Src Addres, Dst Address, Src Port y Dst Port.
5. Estadísticas
a. Muestre todas las IP origen
b. Muestre todas las IP destino
c. ¿Cuál es la IP origen más frecuente?
i. ¿A qué IP destino se comunica con más frecuencia?
ii. ¿A que puerto destino se comunica? ¿Cuál es el propósito de este puerto?


1. Descargue e archivo analisis_paquetes.pcap y asignelo a una variable.

In [ ]:
from scapy.all import rdpcap, IP, TCP
import pandas as pd


paquetes = rdpcap("analisis_paquetes.pcap")


2: Convertir los paquetes a un DataFrame

In [11]:
print("Cantidad de paquetes leídos:", len(paquetes))


Cantidad de paquetes leídos: 62


In [ ]:
from scapy.all import rdpcap, IP, TCP, UDP
import pandas as pd


paquetes = rdpcap("analisis_paquetes.pcap")
datos = []

# Extraer info de cada paquete
for pkt in paquetes:
    if IP in pkt:
        src_ip = pkt[IP].src
        dst_ip = pkt[IP].dst

        # Obtener puerto origen y destino
        if TCP in pkt:
            src_port = pkt[TCP].sport
            dst_port = pkt[TCP].dport
        elif UDP in pkt:
            src_port = pkt[UDP].sport
            dst_port = pkt[UDP].dport
        else:
            src_port = None
            dst_port = None

        datos.append([src_ip, dst_ip, src_port, dst_port])


df = pd.DataFrame(datos, columns=['Src Address', 'Dst Address', 'Src Port', 'Dst Port'])


print(df.head())



   Src Address  Dst Address  Src Port  Dst Port
0   10.1.10.53  84.54.22.33        53        53
1  84.54.22.33   10.1.10.53        53        53
2   10.1.10.53  84.54.22.33        53        53
3  84.54.22.33   10.1.10.53        53        53
4   10.1.10.53  84.54.22.33        53        53


3. Muestre el contenido de las primeras 5 filas del dataset.

In [17]:
print(df.head())


   Src Address  Dst Address  Src Port  Dst Port
0   10.1.10.53  84.54.22.33        53        53
1  84.54.22.33   10.1.10.53        53        53
2   10.1.10.53  84.54.22.33        53        53
3  84.54.22.33   10.1.10.53        53        53
4   10.1.10.53  84.54.22.33        53        53


4. Muestre los valores de las columnas: Src Addres, Dst Address, Src Port y Dst Port.

In [18]:
print(df[['Src Address', 'Dst Address', 'Src Port', 'Dst Port']])


    Src Address  Dst Address  Src Port  Dst Port
0    10.1.10.53  84.54.22.33        53        53
1   84.54.22.33   10.1.10.53        53        53
2    10.1.10.53  84.54.22.33        53        53
3   84.54.22.33   10.1.10.53        53        53
4    10.1.10.53  84.54.22.33        53        53
..          ...          ...       ...       ...
57  84.54.22.33   10.1.10.53        53        53
58   10.1.10.53  84.54.22.33        53        53
59  84.54.22.33   10.1.10.53        53        53
60   10.1.10.53  84.54.22.33        53        53
61  84.54.22.33   10.1.10.53        53        53

[62 rows x 4 columns]


a. Muestre todas las IP origen

In [19]:
print("IPs de origen únicas:")
print(df['Src Address'].unique())


IPs de origen únicas:
['10.1.10.53' '84.54.22.33' '75.75.75.75']


b. Muestre todas las IP destino

In [20]:
print("IPs de destino únicas:")
print(df['Dst Address'].unique())


IPs de destino únicas:
['84.54.22.33' '10.1.10.53' '75.75.75.75']


¿Cuál es la IP origen más frecuente?
i. ¿A qué IP destino se comunica con más frecuencia?
ii. ¿A que puerto destino se comunica? ¿Cuál es el propósito de este puerto?
i. ¿Desde que puertos origen se comunica?

In [21]:
ip_origen_frecuente = df['Src Address'].value_counts().idxmax()
print("IP origen más frecuente:", ip_origen_frecuente)


IP origen más frecuente: 10.1.10.53


In [22]:
ip_destino_frecuente = df[df['Src Address'] == ip_origen_frecuente]['Dst Address'].value_counts().idxmax()
print("IP destino más frecuente para esa IP origen:", ip_destino_frecuente)


IP destino más frecuente para esa IP origen: 84.54.22.33


In [23]:
puerto_destino_frecuente = df[df['Src Address'] == ip_origen_frecuente]['Dst Port'].mode()[0]
print("Puerto destino más frecuente:", puerto_destino_frecuente)


Puerto destino más frecuente: 53


In [24]:
import socket

def obtener_servicio_puerto(puerto):
    try:
        return socket.getservbyport(int(puerto))
    except:
        return "Desconocido"

print("Propósito del puerto destino:", obtener_servicio_puerto(puerto_destino_frecuente))


Propósito del puerto destino: domain


In [25]:
puertos_origen_usados = df[df['Src Address'] == ip_origen_frecuente]['Src Port'].dropna().unique()
print("Puertos de origen utilizados por esa IP:", list(puertos_origen_usados))


Puertos de origen utilizados por esa IP: [np.int64(53), np.int64(15812), np.int64(23903)]


## ¿Cuál es la IP origen más frecuente?	
10.1.10.53

## ¿A qué IP destino se comunica con más frecuencia?	
84.54.22.33

## ¿A qué puerto destino se comunica más frecuentemente?	
53

## ¿Cuál es el propósito de ese puerto?	
domain (es decir, DNS)

## ¿Desde qué puertos origen se comunica?	
[53, 15812, 23903]